### PR TITLE
add support for Dynamic Resource Allocation

### PR DESCRIPTION
**Pull request by moshe010**
Dynamic Resource Allocation was added to k8s 1.26 [1]. 
In k8s 1.27 we added PodResources API to expose the dynamic resources from kubelet.

This PR allow multus to get Dynamic Resource from the podResource API and pass it to CNI as DeviceID.
To use this you will need k8s .127 with DynamicResourceAllocation and  KubeletPodResourcesDynamicResources feature gates enabled. 

[1] - https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation
[2] - https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3695-pod-resources-for-dra

### CODE REVIEW

File: pkg/kubeletclient/kubeletclient.go [hunk 1]
```diff
@@ -144,6 +145,24 @@ func (rc *kubeletClient) GetPodResourceMap(pod *v1.Pod) (map[string]*types.Resou
...
 				}
+				for _, dynamicResource := range cnt.DynamicResources {
+					var deviceIDs []string
+					for _, claimResource := range dynamicResource.ClaimResources {
+						for _, cdiDevice := range claimResource.CDIDevices {
+							res := strings.Split(cdiDevice.Name, "=")
+							if len(res) == 2 {
+								deviceIDs = append(deviceIDs, res[1])
+							} else {
+								logging.Errorf("GetPodResourceMap: Invaid CDI format")
```
Comment (maiqueb) [inline pkg/kubeletclient/kubeletclient.go:156]:
```suggestion
								logging.Errorf("GetPodResourceMap: Invalid CDI format")
```

Comment (moshe010) [reply pkg/kubeletclient/kubeletclient.go]: done

**Comment by bn222** [comment]
Moshe, small nit before we look into this: please split up the commit into "upgrade kube to v0.27.0 + vendor that" and then do the remaining "real" changes. This would make it easier to review.

**Comment by moshe010** [comment]
> Moshe, small nit before we look into this: please split up the commit into "upgrade kube to v0.27.0 + vendor that" and then do the remaining "real" changes. This would make it easier to review.

Sure will do

**Comment by moshe010** [comment]
are there more comments on this PR?

**Comment by bn222** [comment]
The main comment (without looking into how DRA exactly works) from my side is that we might want to have a function that gets or adds the dynamic resources. The current change adds the for-loop in line, which increases the nesting quite a bit.

After reading the code, I want to make sure I understand what's going on here. @moshe010 , is there are reason why we don't want to change pkg/checkpoint/checkpoint.go ? 

Most importantly, @moshe010 , Vrinda from my team will be taking a look into this in more detail. Please allow a bit more time for some comments. We want to understand how this would integrate with sriov network operator.

**Comment by moshe010** [comment]
> The main comment (without looking into how DRA exactly works) from my side is that we might want to have a function that gets or adds the dynamic resources. The current change adds the for-loop in line, which increases the nesting quite a bit.
> 
Sure I will created separate function for dynamic resources
> After reading the code, I want to make sure I understand what's going on here. @moshe010 , is there are reason why we don't want to change pkg/checkpoint/checkpoint.go ?
The checkpoint for multus was introduce before pod resource api exists for device plugin.
For DRA pod resource API will always exist so we don't need to support the fallback. 
Using checkpoint in general is problematic because it may change between k8s version.  
> 
> Most importantly, @moshe010 , Vrinda from my team will be taking a look into this in more detail. Please allow a bit more time for some comments. We want to understand how this would integrate with sriov network operator.

**Comment by wizhaoredhat** [comment]
LGTM

**Comment by vrindle** [comment]
@moshe010 @bn222 I deployed these changes on my custom multus setup. It seems to look good. I deployed these custom changes on my setup and it looks to be good. Since these changes are additive this was expected. It LGTM.

**Comment by moshe010** [comment]
Is there any other who required to review or can we merge this?

**Comment by maiqueb** [comment]
Can you update the docs ? 

I am looking for a couple paragraphs about the feature and some user documentation on how can the user it.

**Comment by moshe010** [comment]
> Can you update the docs ?
> 
> I am looking for a couple paragraphs about the feature and some user documentation on how can the user it.

sure you want to me to add DRA section here https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md?

**Comment by maiqueb** [comment]
> > Can you update the docs ?
> > I am looking for a couple paragraphs about the feature and some user documentation on how can the user it.
> 
> sure, you want to me to add DRA section here https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md?

Unsure what's the best place ... 

But yeah, let's put it there, if needed you'll move it around. 

Asking @s1061123 / @dougbtv 's opinion on this.

**Comment by moshe010** [comment]
The CI failed don't seem to be related to the PR  see [1]. 
@maiqueb sorry for the late response, I updated the doc in  how-to-use.md



[1] - 
Notice:  A new release of pip is available: 23.2.1 -> 23.3.1
Notice:  To update, run: pip install --upgrade pip
Traceback (most recent call last):
  File "/home/runner/.local/bin/j2", line 5, in <module>
    from j2cli import main
  File "/home/runner/.local/lib/python3.12/site-packages/j2cli/__init__.py", line 4, in <module>
    import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'
Error: Process completed with exit code 1.

**Comment by maiqueb** [comment]
> The CI failed don't seem to be related to the PR see [1]. @maiqueb sorry for the late response, I updated the doc in how-to-use.md
> 
> [1] - Notice: A new release of pip is available: 23.2.1 -> 23.3.1 Notice: To update, run: pip install --upgrade pip Traceback (most recent call last): File "/home/runner/.local/bin/j2", line 5, in from j2cli import main File "/home/runner/.local/lib/python3.12/site-packages/j2cli/**init**.py", line 4, in import pkg_resources ModuleNotFoundError: No module named 'pkg_resources' Error: Process completed with exit code 1.

Hey @moshe010 ; all good. 

Regarding jinja, you need to update it to latest version (we actually couldn't find an updated version, and thus replaced it). 
[Here's what was done](https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller/pull/183/files) in the multus dynamic networks repo.

File: docs/how-to-use.md [hunk 2]
```diff
@@ -634,3 +634,116 @@ Sometimes, you may wish to not have the entrypoint copy the binary file onto the
...
+#### Prerequisite
+
+1. Kubernetes 1.27
+2. Container Runtime with CDI support enabled
+3. Kubernetes runtime-config=resource.k8s.io/v1alpha2
+4. Kubernetes feature-gates=DynamicResourceAllocation=True,KubeletPodResourcesDynamicResources=true
+
+#### Install DRA driver
+
+The current example uses Nvidia DRA driver for networking. This DRA driver is not public avaialbe.
```
Comment (maiqueb) [inline docs/how-to-use.md:653]: typo.

```suggestion
The current example uses Nvidia DRA driver for networking. This DRA driver is not publicly available.
```

**Review by maiqueb — COMMENTED** [review]
Neat docs.

Just pointing out some typo.

**Comment by moshe010** [comment]
> > The CI failed don't seem to be related to the PR see [1]. @maiqueb sorry for the late response, I updated the doc in how-to-use.md
> > [1] - Notice: A new release of pip is available: 23.2.1 -> 23.3.1 Notice: To update, run: pip install --upgrade pip Traceback (most recent call last): File "/home/runner/.local/bin/j2", line 5, in from j2cli import main File "/home/runner/.local/lib/python3.12/site-packages/j2cli/**init**.py", line 4, in import pkg_resources ModuleNotFoundError: No module named 'pkg_resources' Error: Process completed with exit code 1.
> 
> Hey @moshe010 ; all good.
> 
> Regarding jinja, you need to update it to latest version (we actually couldn't find an updated version, and thus replaced it). [Here's what was done](https://github.com/k8snetworkplumbingwg/multus-dynamic-networks-controller/pull/183/files) in the multus dynamic networks repo.

I see so it mean that some will need to do similar fix to this repo to fix the jinja error.

**Comment by moshe010** [comment]
Are we ok with this change?  are we waiting for someone else to review it? 
we will really appreciate help in review it getting merged to multus.

**Review by s1061123 — CHANGES_REQUESTED** [review]

**Comment by vasrem** [comment]
@s1061123 @dougbtv:

Folks, I've added an integration test with an example dra driver implementation as discussed in the last community meeting. Please have a look at the latest commit and let me if that's sufficient to get that PR merged.

Notice the condition on running the integration test. This looks irrelevant to this PR. If you have any insights let me know.

**Comment by vasrem** [comment]
@s1061123 @dougbtv:

We now run the e2e tests for both thick and thin as requested in the community meeting today. The difference is that all the DRA example Pods are running on the worker nodes. Before, the controller Pod was running on the control plane node where the network is not stable. I have added this PR https://github.com/k8snetworkplumbingwg/multus-cni/pull/1259 to demonstrate what's failing and prove that this is not a regression introduced by this PR.

File: docs/how-to-use.md [hunk 3]
```diff
@@ -634,3 +634,116 @@ Sometimes, you may wish to not have the entrypoint copy the binary file onto the
 If you wish to have auto configuration use the `readinessindicatorfile` in the configuration, you can use the `--readiness-indicator-file` to express which file should be used as the readiness indicator.
 
     --readiness-indicator-file=/path/to/file
+
+### Run pod with network annotation and Dynamic Resource Allocation driver
+
+Dynamic Resource Allocation is alternative mechanism to device plugin which allow to requests pod and container resources. The feature is alpha in k8s 1.27.
+
+The following sections describe how to use DRA with multus and Nvidia DRA driver. Other DRA networking driver vendors should follow similar concepts to make use of multus DRA support.
+
```
Comment (dougbtv) [inline docs/how-to-use.md:643]: Can we add some text along the lines of:

```
Dynamic Resource Allocation (DRA) is [currently an alpha](https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/), and is subject to change. Please consider this functionality as a preview, and utilized a potentially replace for device plugin for SR-IOV. The architecture and usage of DRA in Multus CNI may be changed in the future as this technology matures.
```

**Review by dougbtv — CHANGES_REQUESTED** [review]

File: docs/how-to-use.md (hunk 3 above)
Comment (vasrem) [reply docs/how-to-use.md:649]: Thank you, updated! Added the fact that DRA is in alpha as warning.

**Pull request merged**